index.js

```diff
@@ -118,8 +118,95 @@ async function getAllDatasets(dataCatalogUrl = 'https://openactive.io/data-catal
+ *
+ * This function performs an HTTP GET request to the specified ID (URL), retrieves
+ * the response, and extracts JSON-LD from it if needed and possible. It then compares
+ * the '@id' or 'id' field from the retrieved JSON-LD to the provided ID.
+ *
+ * Note that this is only applicable to JSON-LD "@id" for the DataCatalog and Dataset types, which must resolve.
+ *
+ * @async
+ * @function validateJsonLdId
+ * @param {string} id - The expected '@id' or 'id' value, also the URL to be requested.
+ * @param {boolean} expectHtml - A flag indicating whether the response is expected to be HTML (i.e. a Dataset Site).
+ * @returns {Promise<{isValid: boolean, error: string|null}>} - An object indicating the validity
+ * of the JSON-LD and any associated error message.
+ *
+ * @example
+ * validateJsonLdId('https://example.com/data.jsonld', false)
+ *   .then(({isValid, error}) => {
+ *     if (isValid) {
+ *       console.log('JSON-LD is valid!');
+ *     } else {
+ *       console.error(`JSON-LD validation failed: ${error}`);
+ *     }
+ *   });
+ */
+async function validateJsonLdId(id, expectHtml) {
+  let response;
+
+  try {
+    response = await axiosGetWithRetry(id);
+    response = response.data;
+  } catch (error) {
+    return { isValid: false, error: `Failed to resolve URL: ${error.message}` };
+  }
+
+  let jsonLd;
+  try {
+    if (expectHtml && typeof response === 'string') {
+      jsonLd = extractJSONLDfromHTML(id, response);
+    } else if (!expectHtml && typeof response === 'object') {
+      jsonLd = response;
+    } else {
+      return { isValid: false, error: `Unexpected response type: ${typeof response}` };
+    }
+
+    const jsonId = jsonLd['@id'] || jsonLd.id;
+    if (jsonId !== id) {
+      return { isValid: false, error: `Mismatched '@id': From file: "${id}"; From referenced JSON-LD: "${jsonId}"` };
+    }
+  } catch (error) {
+    return { isValid: false, error: error.message };
+  }
+
+  return { isValid: true, error: null };
+}
+
+/*
+* System-specific workaround: Note that rate limits in Legend can cause this request to fail with a 403 (?), so we retry up to 5 times
+* TODO: Ask Legend to return a 429 instead
+*/
+async function axiosGetWithRetry(url) {
+  let response;
+  const maxRetries = 5; // Define a maximum number of retries
+
+  async function sleep(milliseconds) {
+    return new Promise((resolve) => { setTimeout(resolve, milliseconds); });
+  }
+
+  for (let attempt = 0; attempt < maxRetries; attempt += 1) {
+    try {
+      response = await axios.get(url);
+      break; // Exit the loop if the request was successful
+    } catch (error) {
+      if (error.response && error.response.status === 403 && attempt < maxRetries - 1) {
+        // Log a warning and retry after sleeping for a random duration between 1 and 3 seconds
+        console.warn(`Attempt ${attempt + 1}: Access forbidden (403) for URL: ${url}. Retrying...`);
+        await sleep(1000 + Math.random() * 2000); // Sleep for 1 to 3 seconds
```

Comment:
Why a random duration as opposed to e.g. exponential retry? If this still makes sense, would you be able to update the code with the reason?

Reply:
It's so it doesn't overwhelm a server that's serving multiple datasets if multiple instances of this function are run in parallel - will document